File: libs/agentcard-go/agentcard.go
package agentcard

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// DID represents a Decentralized Identifier
type DID string

// NewAgentDID creates a new agent DID with format: did:ainur:agent:<identifier>
func NewAgentDID(identifier string) DID {
	return DID(fmt.Sprintf("did:ainur:agent:%s", identifier))
}

// NewUserDID creates a new user DID with format: did:ainur:user:<identifier>
func NewUserDID(identifier string) DID {
	return DID(fmt.Sprintf("did:ainur:user:%s", identifier))
}

// NewNetworkDID creates a new network DID
func NewNetworkDID(identifier string) DID {
	return DID(fmt.Sprintf("did:ainur:network:%s", identifier))
}

// String returns the DID as a string
func (d DID) String() string {
	return string(d)
}

// Operation defines a capability operation
type Operation struct {
	Name         string                 `json:"name"`
	Category     string                 `json:"category"`
	InputSchema  map[string]interface{} `json:"input_schema,omitempty"`
	OutputSchema map[string]interface{} `json:"output_schema,omitempty"`
	Complexity   string                 `json:"complexity,omitempty"`
	GasEstimate  uint64                 `json:"gas_estimate"`
}

// CapabilityConstraints defines resource limits
type CapabilityConstraints struct {
	MaxInputSize       uint64  `json:"max_input_size"`
	MaxExecutionTimeMs uint64  `json:"max_execution_time_ms"`
	ConcurrentTasks    *uint32 `json:"concurrent_tasks,omitempty"`
}

// Capabilities declares what an agent can do
type Capabilities struct {
	Domains     []string              `json:"domains"`
	Operations  []Operation           `json:"operations"`
	Constraints CapabilityConstraints `json:"constraints"`
	Interfaces  []string              `json:"interfaces"`
}

// ExecutionEnvironment defines runtime constraints
type ExecutionEnvironment struct {
	MemoryLimitMB     uint32 `json:"memory_limit_mb"`
	CPUQuotaMs        uint32 `json:"cpu_quota_ms"`
	NetworkEnabled    bool   `json:"network_enabled"`
	FilesystemEnabled bool   `json:"filesystem_enabled"`
}

// Endpoint defines a network endpoint
type Endpoint struct {
	Protocol string `json:"protocol"`
	Address  string `json:"address"`
	TLS      *bool  `json:"tls,omitempty"`
}

// RuntimeInfo describes the agent's runtime environment
type RuntimeInfo struct {
	Protocol             string               `json:"protocol"`
	Implementation       string               `json:"implementation"`
	Version              string               `json:"version"`
	WasmEngine           string               `json:"wasm_engine"`
	WasmVersion          string               `json:"wasm_version"`
	ModuleHash           string               `json:"module_hash"`
	ModuleURL            string               `json:"module_url,omitempty"`
	ExecutionEnvironment ExecutionEnvironment `json:"execution_environment"`
	Endpoints            []Endpoint           `json:"endpoints"`
}

// Badge represents a reputation badge
type Badge struct {
	Type      string    `json:"type"`
	Threshold string    `json:"threshold,omitempty"`
	IssuedBy  DID       `json:"issued_by"`
	IssuedAt  time.Time `json:"issued_at"`
}

// Reputation tracks agent trustworthiness
type Reputation struct {
	TrustScore             float64                  `json:"trust_score"`
	TotalTasks             uint64                   `json:"total_tasks"`
	SuccessfulTasks        uint64                   `json:"successful_tasks"`
	FailedTasks            uint64                   `json:"failed_tasks"`
	SuccessRate            float64                  `json:"success_rate"`
	AverageExecutionTimeMs uint64                   `json:"average_execution_time_ms"`
	UptimePercentage       float64                  `json:"uptime_percentage"`
	PeerEndorsements       uint32                   `json:"peer_endorsements"`
	Violations             uint32                   `json:"violations"`
	CreatedAt              time.Time                `json:"created_at"`
	LastActive             time.Time                `json:"last_active"`
	Badges                 []Badge                  `json:"badges"`
	SlashingHistory        []map[string]interface{} `json:"slashing_history"`
}

// NewReputation creates a default reputation for a new agent
func NewReputation() Reputation {
	now := time.Now()
	return Reputation{
		TrustScore:       50.0,
		TotalTasks:       0,
		SuccessfulTasks:  0,
		FailedTasks:      0,
		SuccessRate:      0.0,
		UptimePercentage: 100.0,
		CreatedAt:        now,
		LastActive:       now,
		Badges:           []Badge{},
		SlashingHistory:  []map[string]interface{}{},
	}
}

// Discount defines a pricing discount
type Discount struct {
	Type               string   `json:"type"`
	MinTasks           *uint64  `json:"min_tasks,omitempty"`
	MinTrustScore      *float64 `json:"min_trust_score,omitempty"`
	DiscountPercentage float64  `json:"discount_percentage"`
}

// SurgePricing defines dynamic pricing
type SurgePricing struct {
	Enabled         bool    `json:"enabled"`
	MultiplierMax   float64 `json:"multiplier_max"`
	DemandThreshold float64 `json:"demand_threshold"`
}

// Economic defines pricing and payment
type Economic struct {
	PricingModel    string        `json:"pricing_model"`
	BasePriceUAinur uint64        `json:"base_price_uainur"`
	SurgePricing    *SurgePricing `json:"surge_pricing,omitempty"`
	Discounts       []Discount    `json:"discounts"`
	PaymentMethods  []string      `json:"payment_methods"`
	EscrowRequired  bool          `json:"escrow_required"`
	RefundPolicy    string        `json:"refund_policy"`
}

// NewEconomic creates default economic parameters
func NewEconomic() Economic {
	return Economic{
		PricingModel:    "per_operation",
		BasePriceUAinur: 100,
		Discounts:       []Discount{},
		PaymentMethods:  []string{"ainur"},
		EscrowRequired:  false,
		RefundPolicy:    "full_refund_on_failure",
	}
}

// P2PConfig defines P2P network settings
type P2PConfig struct {
	PeerID            string   `json:"peer_id"`
	ListenAddresses   []string `json:"listen_addresses"`
	AnnounceAddresses []string `json:"announce_addresses"`
	Protocols         []string `json:"protocols"`
}

// Discovery defines discovery methods
type Discovery struct {
	Methods        []string `json:"methods"`
	BootstrapNodes []string `json:"bootstrap_nodes"`
}

// LatencyTargets defines expected latency
type LatencyTargets struct {
	P50Ms uint64 `json:"p50_ms"`
	P95Ms uint64 `json:"p95_ms"`
	P99Ms uint64 `json:"p99_ms"`
}

// Availability defines where the agent is available
type Availability struct {
	Regions        []string       `json:"regions"`
	LatencyTargets LatencyTargets `json:"latency_targets"`
}

// Network defines networking configuration
type Network struct {
	P2P          P2PConfig    `json:"p2p"`
	Discovery    Discovery    `json:"discovery"`
	Availability Availability `json:"availability"`
}

// CredentialSubject is the core agent data
type CredentialSubject struct {
	ID           DID          `json:"id"`
	Type         string       `json:"type"`
	Name         string       `json:"name"`
	Description  string       `json:"description"`
	Version      string       `json:"version"`
	Capabilities Capabilities `json:"capabilities"`
	Runtime      RuntimeInfo  `json:"runtime"`
	Reputation   Reputation   `json:"reputation"`
	Economic     Economic     `json:"economic"`
	Network      Network      `json:"network"`
}

// Proof is the cryptographic signature
type Proof struct {
	Type               string    `json:"type"`
	Created            time.Time `json:"created"`
	VerificationMethod string    `json:"verification_method"`
	ProofPurpose       string    `json:"proof_purpose"`
	ProofValue         string    `json:"proof_value"`
}

// AgentCard is a W3C Verifiable Credential for agents
type AgentCard struct {
	Context           []string          `json:"@context"`
	ID                string            `json:"id"`
	Type              []string          `json:"type"`
	Issuer            DID               `json:"issuer"`
	IssuanceDate      time.Time         `json:"issuanceDate"`
	ExpirationDate    time.Time         `json:"expirationDate"`
	CredentialSubject CredentialSubject `json:"credentialSubject"`
	Proof             *Proof            `json:"proof,omitempty"`
}

// AgentCardBuilder helps construct AgentCards
type AgentCardBuilder struct {
	agentDID       *DID
	name           string
	description    string
	version        string
	capabilities   *Capabilities
	runtime        *RuntimeInfo
	reputation     *Reputation
	economic       *Economic
	network        *Network
	issuer         *DID
	expirationDays uint64
}

// NewAgentCardBuilder creates a new builder
func NewAgentCardBuilder() *AgentCardBuilder {
	return &AgentCardBuilder{
		expirationDays: 365,
	}
}

// SetAgentDID sets the agent's DID
func (b *AgentCardBuilder) SetAgentDID(did DID) *AgentCardBuilder {
	b.agentDID = &did
	return b
}

// SetName sets the agent's name
func (b *AgentCardBuilder) SetName(name string) *AgentCardBuilder {
	b.name = name
	return b
}

// SetDescription sets the agent's description
func (b *AgentCardBuilder) SetDescription(desc string) *AgentCardBuilder {
	b.description = desc
	return b
}

// SetVersion sets the agent's version
func (b *AgentCardBuilder) SetVersion(version string) *AgentCardBuilder {
	b.version = version
	return b
}

// SetCapabilities sets the agent's capabilities
func (b *AgentCardBuilder) SetCapabilities(cap Capabilities) *AgentCardBuilder {
	b.capabilities = &cap
	return b
}

// SetRuntime sets the runtime information
func (b *AgentCardBuilder) SetRuntime(runtime RuntimeInfo) *AgentCardBuilder {
	b.runtime = &runtime
	return b
}

// SetReputation sets the reputation
func (b *AgentCardBuilder) SetReputation(rep Reputation) *AgentCardBuilder {
	b.reputation = &rep
	return b
}

// SetEconomic sets economic parameters
func (b *AgentCardBuilder) SetEconomic(econ Economic) *AgentCardBuilder {
	b.economic = &econ
	return b
}

// SetNetwork sets network configuration
func (b *AgentCardBuilder) SetNetwork(net Network) *AgentCardBuilder {
	b.network = &net
	return b
}

// SetIssuer sets the issuer DID
func (b *AgentCardBuilder) SetIssuer(issuer DID) *AgentCardBuilder {
	b.issuer = &issuer
	return b
}

// SetExpirationDays sets expiration in days
func (b *AgentCardBuilder) SetExpirationDays(days uint64) *AgentCardBuilder {
	b.expirationDays = days
	return b
}

// Build constructs the AgentCard
func (b *AgentCardBuilder) Build() (*AgentCard, error) {
	if b.agentDID == nil {
		return nil, fmt.Errorf("agent_did is required")
	}
	if b.name == "" {
		return nil, fmt.Errorf("name is required")
	}
	if b.description == "" {
		return nil, fmt.Errorf("description is required")
	}
	if b.capabilities == nil {
		return nil, fmt.Errorf("capabilities are required")
	}
	if b.runtime == nil {
		return nil, fmt.Errorf("runtime is required")
	}
	if b.network == nil {
		return nil, fmt.Errorf("network is required")
	}

	version := b.version
	if version == "" {
		version = "1.0.0"
	}

	reputation := b.reputation
	if reputation == nil {
		rep := NewReputation()
		reputation = &rep
	}

	economic := b.economic
	if economic == nil {
		econ := NewEconomic()
		economic = &econ
	}

	issuer := b.issuer
	if issuer == nil {
		issuer = b.agentDID
	}

	now := time.Now()
	expiration := now.AddDate(0, 0, int(b.expirationDays))
	cardID := fmt.Sprintf("did:ainur:agentcard:%s", uuid.New().String())

	return &AgentCard{
		Context: []string{
			"https://www.w3.org/2018/credentials/v1",
			"https://ainur.network/contexts/agentcard/v1",
		},
		ID:             cardID,
		Type:           []string{"VerifiableCredential", "AgentCard"},
		Issuer:         *issuer,
		IssuanceDate:   now,
		ExpirationDate: expiration,
		CredentialSubject: CredentialSubject{
			ID:           *b.agentDID,
			Type:         "AutonomousAgent",
			Name:         b.name,
			Description:  b.description,
			Version:      version,
			Capabilities: *b.capabilities,
			Runtime:      *b.runtime,
			Reputation:   *reputation,
			Economic:     *economic,
			Network:      *b.network,
		},
		Proof: nil,
	}, nil
}

// ToJSON converts the AgentCard to JSON
func (ac *AgentCard) ToJSON() ([]byte, error) {
	return json.MarshalIndent(ac, "", "  ")
}

// FromJSON parses an AgentCard from JSON
func FromJSON(data []byte) (*AgentCard, error) {
	var card AgentCard
	if err := json.Unmarshal(data, &card); err != nil {
		return nil, err
	}
	return &card, nil
}

// Sign signs the AgentCard with an Ed25519 private key
func (ac *AgentCard) Sign(privateKey ed25519.PrivateKey) error {
	// Remove any existing proof
	ac.Proof = nil

	// Create canonical JSON of credential subject
	subjectJSON, err := json.Marshal(ac.CredentialSubject)
	if err != nil {
		return fmt.Errorf("failed to marshal credential subject: %w", err)
	}

	// Sign the canonical JSON
	signature := ed25519.Sign(privateKey, subjectJSON)

	// Encode signature as base64
	proofValue := base64.StdEncoding.EncodeToString(signature)

	// Create verification method DID
	verificationMethod := fmt.Sprintf("%s#keys-1", ac.CredentialSubject.ID)

	// Add proof
	ac.Proof = &Proof{
		Type:               "Ed25519Signature2020",
		Created:            time.Now(),
		VerificationMethod: verificationMethod,
		ProofPurpose:       "assertionMethod",
		ProofValue:         proofValue,
	}

	return nil
}

// Verify verifies the AgentCard signature
func (ac *AgentCard) Verify(publicKey ed25519.PublicKey) (bool, error) {
	if ac.Proof == nil {
		return false, fmt.Errorf("agentcard has no proof")
	}

	// Decode signature from base64
	signature, err := base64.StdEncoding.DecodeString(ac.Proof.ProofValue)
	if err != nil {
		return false, fmt.Errorf("failed to decode signature: %w", err)
	}

	// Recreate canonical JSON (without proof)
	proofBackup := ac.Proof
	ac.Proof = nil
	subjectJSON, err := json.Marshal(ac.CredentialSubject)
	ac.Proof = proofBackup
	if err != nil {
		return false, fmt.Errorf("failed to marshal credential subject: %w", err)
	}

	// Verify signature
	return ed25519.Verify(publicKey, subjectJSON, signature), nil
}

// Hash creates a SHA-256 hash of the AgentCard
func (ac *AgentCard) Hash() (string, error) {
	subjectJSON, err := json.Marshal(ac.CredentialSubject)
	if err != nil {
		return "", err
	}

	hash := sha256.Sum256(subjectJSON)
	return fmt.Sprintf("sha256:%s", hex.EncodeToString(hash[:])), nil
}

// GenerateKeyPair generates a new Ed25519 keypair
func GenerateKeyPair() (ed25519.PublicKey, ed25519.PrivateKey, error) {
	return ed25519.GenerateKey(rand.Reader)
}

File: examples/python-sdk/agentcard.py
"""
AgentCard-VC-v1 Python SDK

Simple Python implementation for creating and working with AgentCards
following the W3C Verifiable Credentials standard.
"""

import json
import hashlib
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, field, asdict
from enum import Enum


class DIDType(Enum):
    AGENT = "agent"
    USER = "user"
    NETWORK = "network"


@dataclass
class DID:
    """Decentralized Identifier"""
    identifier: str
    did_type: DIDType
    
    def __str__(self) -> str:
        return f"did:ainur:{self.did_type.value}:{self.identifier}"
    
    @classmethod
    def agent(cls, identifier: str) -> 'DID':
        return cls(identifier, DIDType.AGENT)
    
    @classmethod
    def user(cls, identifier: str) -> 'DID':
        return cls(identifier, DIDType.USER)
    
    @classmethod
    def network(cls, identifier: str) -> 'DID':
        return cls(identifier, DIDType.NETWORK)


@dataclass
class Operation:
    """Agent capability operation"""
    name: str
    category: str
    gas_estimate: int = 100
    input_schema: Optional[Dict[str, Any]] = None
    output_schema: Optional[Dict[str, Any]] = None
    complexity: Optional[str] = None


@dataclass
class CapabilityConstraints:
    """Resource constraints for capabilities"""
    max_input_size: int = 1048576  # 1MB
    max_execution_time_ms: int = 30000  # 30 seconds
    concurrent_tasks: Optional[int] = None


@dataclass
class Capabilities:
    """Agent capabilities"""
    domains: List[str]
    operations: List[Operation]
    constraints: CapabilityConstraints
    interfaces: List[str] = field(default_factory=lambda: ["http", "grpc"])


@dataclass
class Badge:
    """Reputation badge"""
    badge_type: str
    threshold: Optional[str] = None
    issued_by: Optional[str] = None
    issued_at: Optional[str] = None


@dataclass
class Reputation:
    """Agent reputation metrics"""
    trust_score: float = 50.0
    total_tasks: int = 0
    successful_tasks: int = 0
    failed_tasks: int = 0
    success_rate: float = 0.0
    average_execution_time_ms: int = 0
    uptime_percentage: float = 100.0
    peer_endorsements: int = 0
    violations: int = 0
    created_at: str = field(default_factory=lambda: datetime.utcnow().isoformat() + "Z")
    last_active: str = field(default_factory=lambda: datetime.utcnow().isoformat() + "Z")
    badges: List[Badge] = field(default_factory=list)
    slashing_history: List[Dict[str, Any]] = field(default_factory=list)


@dataclass
class Discount:
    """Pricing discount"""
    discount_type: str
    discount_percentage: float
    min_tasks: Optional[int] = None
    min_trust_score: Optional[float] = None


@dataclass
class SurgePricing:
    """Dynamic surge pricing"""
    enabled: bool = False
    multiplier_max: float = 2.0
    demand_threshold: float = 0.8


@dataclass
class Economic:
    """Economic parameters"""
    pricing_model: str = "per_operation"
    base_price_uainur: int = 100
    surge_pricing: Optional[SurgePricing] = None
    discounts: List[Discount] = field(default_factory=list)
    payment_methods: List[str] = field(default_factory=lambda: ["ainur"])
    escrow_required: bool = False
    refund_policy: str = "full_refund_on_failure"


@dataclass
class ExecutionEnvironment:
    """Runtime execution environment"""
    memory_limit_mb: int = 128
    cpu_quota_ms: int = 1000
    network_enabled: bool = True
    filesystem_enabled: bool = False


@dataclass
class Endpoint:
    """Network endpoint"""
    protocol: str
    address: str
    tls: Optional[bool] = None


@dataclass
class RuntimeInfo:
    """Agent runtime information"""
    protocol: str
    implementation: str
    version: str
    wasm_engine: str
    wasm_version: str
    module_hash: str
    module_url: Optional[str] = None
    execution_environment: ExecutionEnvironment = field(default_factory=ExecutionEnvironment)
    endpoints: List[Endpoint] = field(default_factory=list)


@dataclass
class P2PConfig:
    """P2P network configuration"""
    peer_id: str
    listen_addresses: List[str]
    announce_addresses: List[str] = field(default_factory=list)
    protocols: List[str] = field(default_factory=lambda: ["/ainur/gossipsub/1.0.0"])


@dataclass
class Discovery:
    """Discovery configuration"""
    methods: List[str] = field(default_factory=lambda: ["mdns", "dht"])
    bootstrap_nodes: List[str] = field(default_factory=list)


@dataclass
class LatencyTargets:
    """Network latency targets"""
    p50_ms: int = 50
    p95_ms: int = 100
    p99_ms: int = 200


@dataclass
class Availability:
    """Availability configuration"""
    regions: List[str] = field(default_factory=lambda: ["global"])
    latency_targets: LatencyTargets = field(default_factory=LatencyTargets)


@dataclass
class Network:
    """Network configuration"""
    p2p: P2PConfig
    discovery: Discovery
    availability: Availability = field(default_factory=Availability)


@dataclass
class CredentialSubject:
    """The core agent data"""
    id: str
    subject_type: str
    name: str
    description: str
    version: str
    capabilities: Capabilities
    runtime: RuntimeInfo
    reputation: Reputation
    economic: Economic
    network: Network


@dataclass
class Proof:
    """Cryptographic proof"""
    proof_type: str
    created: str
    verification_method: str
    proof_purpose: str
    proof_value: str


@dataclass
class AgentCard:
    """W3C Verifiable Credential for agent identity"""
    context: List[str]
    id: str
    card_type: List[str]
    issuer: str
    issuance_date: str
    expiration_date: str
    credential_subject: CredentialSubject
    proof: Optional[Proof] = None
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary with proper JSON-LD keys"""
        data = {
            "@context": self.context,
            "id": self.id,
            "type": self.card_type,
            "issuer": self.issuer,
            "issuanceDate": self.issuance_date,
            "expirationDate": self.expiration_date,
            "credentialSubject": self._serialize_subject(self.credential_subject)
        }
        if self.proof:
            data["proof"] = asdict(self.proof)
        return data
    
    def _serialize_subject(self, subject: CredentialSubject) -> Dict[str, Any]:
        """Serialize credential subject with proper structure"""
        return {
            "id": subject.id,
            "type": subject.subject_type,
            "name": subject.name,
            "description": subject.description,
            "version": subject.version,
            "capabilities": {
                "domains": subject.capabilities.domains,
                "operations": [asdict(op) for op in subject.capabilities.operations],
                "constraints": asdict(subject.capabilities.constraints),
                "interfaces": subject.capabilities.interfaces
            },
            "runtime": self._serialize_runtime(subject.runtime),
            "reputation": asdict(subject.reputation),
            "economic": self._serialize_economic(subject.economic),
            "network": self._serialize_network(subject.network)
        }
    
    def _serialize_runtime(self, runtime: RuntimeInfo) -> Dict[str, Any]:
        """Serialize runtime info"""
        data = {
            "protocol": runtime.protocol,
            "implementation": runtime.implementation,
            "version": runtime.version,
            "wasm_engine": runtime.wasm_engine,
            "wasm_version": runtime.wasm_version,
            "module_hash": runtime.module_hash,
            "execution_environment": asdict(runtime.execution_environment),
            "endpoints": [asdict(ep) for ep in runtime.endpoints]
        }
        if runtime.module_url:
            data["module_url"] = runtime.module_url
        return data
    
    def _serialize_economic(self, economic: Economic) -> Dict[str, Any]:
        """Serialize economic parameters"""
        data = {
            "pricing_model": economic.pricing_model,
            "base_price_uainur": economic.base_price_uainur,
            "discounts": [asdict(d) for d in economic.discounts],
            "payment_methods": economic.payment_methods,
            "escrow_required": economic.escrow_required,
            "refund_policy": economic.refund_policy
        }
        if economic.surge_pricing:
            data["surge_pricing"] = asdict(economic.surge_pricing)
        return data
    
    def _serialize_network(self, network: Network) -> Dict[str, Any]:
        """Serialize network configuration"""
        return {
            "p2p": asdict(network.p2p),
            "discovery": asdict(network.discovery),
            "availability": {
                "regions": network.availability.regions,
                "latency_targets": asdict(network.availability.latency_targets)
            }
        }
    
    def to_json(self, indent: int = 2) -> str:
        """Convert to JSON string"""
        return json.dumps(self.to_dict(), indent=indent)
    
    def hash(self) -> str:
        """Calculate SHA-256 hash of the credential subject"""
        subject_json = json.dumps(self._serialize_subject(self.credential_subject), sort_keys=True)
        hash_bytes = hashlib.sha256(subject_json.encode()).hexdigest()
        return f"sha256:{hash_bytes}"


class AgentCardBuilder:
    """Builder for creating AgentCards"""
    
    def __init__(self):
        self.agent_did: Optional[DID] = None
        self.name: Optional[str] = None
        self.description: Optional[str] = None
        self.version: str = "1.0.0"
        self.capabilities: Optional[Capabilities] = None
        self.runtime: Optional[RuntimeInfo] = None
        self.reputation: Optional[Reputation] = None
        self.economic: Optional[Economic] = None
        self.network: Optional[Network] = None
        self.issuer: Optional[DID] = None
        self.expiration_days: int = 365
    
    def set_agent_did(self, did: DID) -> 'AgentCardBuilder':
        """Set the agent DID"""
        self.agent_did = did
        return self
    
    def set_name(self, name: str) -> 'AgentCardBuilder':
        """Set the agent name"""
        self.name = name
        return self
    
    def set_description(self, description: str) -> 'AgentCardBuilder':
        """Set the agent description"""
        self.description = description
        return self
    
    def set_version(self, version: str) -> 'AgentCardBuilder':
        """Set the agent version"""
        self.version = version
        return self
    
    def set_capabilities(self, capabilities: Capabilities) -> 'AgentCardBuilder':
        """Set the agent capabilities"""
        self.capabilities = capabilities
        return self
    
    def set_runtime(self, runtime: RuntimeInfo) -> 'AgentCardBuilder':
        """Set the runtime info"""
        self.runtime = runtime
        return self
    
    def set_reputation(self, reputation: Reputation) -> 'AgentCardBuilder':
        """Set the reputation"""
        self.reputation = reputation
        return self
    
    def set_economic(self, economic: Economic) -> 'AgentCardBuilder':
        """Set economic parameters"""
        self.economic = economic
        return self
    
    def set_network(self, network: Network) -> 'AgentCardBuilder':
        """Set network configuration"""
        self.network = network
        return self
    
    def set_issuer(self, issuer: DID) -> 'AgentCardBuilder':
        """Set the issuer DID"""
        self.issuer = issuer
        return self
    
    def set_expiration_days(self, days: int) -> 'AgentCardBuilder':
        """Set expiration in days"""
        self.expiration_days = days
        return self
    
    def build(self) -> AgentCard:
        """Build the AgentCard"""
        if not self.agent_did:
            raise ValueError("agent_did is required")
        if not self.name:
            raise ValueError("name is required")
        if not self.description:
            raise ValueError("description is required")
        if not self.capabilities:
            raise ValueError("capabilities are required")
        if not self.runtime:
            raise ValueError("runtime is required")
        if not self.network:
            raise ValueError("network is required")
        
        # Use defaults
        reputation = self.reputation or Reputation()
        economic = self.economic or Economic()
        issuer = self.issuer or self.agent_did
        
        # Generate timestamps
        now = datetime.utcnow()
        issuance_date = now.isoformat() + "Z"
        expiration_date = (now + timedelta(days=self.expiration_days)).isoformat() + "Z"
        
        # Generate card ID
        import uuid
        card_id = f"did:ainur:agentcard:{uuid.uuid4()}"
        
        # Build credential subject
        subject = CredentialSubject(
            id=str(self.agent_did),
            subject_type="AutonomousAgent",
            name=self.name,
            description=self.description,
            version=self.version,
            capabilities=self.capabilities,
            runtime=self.runtime,
            reputation=reputation,
            economic=economic,
            network=self.network
        )
        
        # Create AgentCard
        return AgentCard(
            context=[
                "https://www.w3.org/2018/credentials/v1",
                "https://ainur.network/contexts/agentcard/v1"
            ],
            id=card_id,
            card_type=["VerifiableCredential", "AgentCard"],
            issuer=str(issuer),
            issuance_date=issuance_date,
            expiration_date=expiration_date,
            credential_subject=subject,
            proof=None
        )
